[DWARF] Emit a worst-case prologue_end flag for pathological inputs

jmorse opened the pull request:
Hi,

(Maybe this should be in an RFC if it's properly contentious,)

Problem: sometimes the entry blocks of functions contain no source-locations at all. Currently, prologue_end is attached to the first source-location found when iterating over the function, which don't have to be in the entry block or anywhere near it. It can even be the return instruction. IMO: this isn't beneficial and can be misleading, the spec says it's "where execution should be suspended for a breakpoint at the entry of a function". If the prologue_end point is past meaningful computation, or past the entry block, the developer has probably missed something useful.

These situations do pop up occasionally: I've got a few game code samples that have been heavily LTO'd, and a non-trivial number of functions end up with a) a lot of information loss after inlining and de-duplication, and b) a prologue_end set after various memory operations and conditional branches. Having to debug LTO'd binaries is already a trial, making it harder to set effective breakpoints doesn't help the developer.

Thus, I'd like to merge this patch which, if there aren't any source locations in the entry block, picks the first non-trivial instruction to apply the scope-line and prologue_end to as a "worst case" or "backup" entry point. See the added dbg-prolog-end-backup-loc.ll for an illustrated example.

However, nothing is ever that simple and I've found myself in the tar pit here, with numerous existing behaviours to support:
 * In un-optimised code it's desirable for prologue_end to fall-through into a non-entry block see the additions to DebugInfo/X86/dbg-prolog-end.ll
 * In some circumstances there are no non-frame-setup instructions in the entry block anyway, in which case I fall back to not emitting a prologue_end.

In terms of test effects:
 * pseudo_cmov_lower2.ll grows a prologue_end at the start of this otherwise location-less function.
 * empty-inline.mir shifts prologue_end upwards into the entry block, before control-flow happens, which is a desirable move IMO.
 * empty-line-info.ll grows prologue_ends on the scope-line linetable entries -- these are entries we add for functions with no source locations anyway.

A test that can't be fixed right now is CodeGen/Thumb2/pr52817.ll, where we put prologue_end half way through substantive computation. This is because it looks a lot like an unoptimized piece of code where prologue_end should live on the first source-location after falling-out of the entry block.

I fed this to the GDB test suite; there weren't any further regressions past the ~121 test failures I had before this patch.

jmorse committed:
Emit a worst-case prologue_end flag for pathological inputs

prologue_end usually indicates where the end of the function-initialization
lies, and is where debuggers usually choose to put the initial breakpoint
for a function. Our current algorithm piggy-backs it on the first available
source-location: which doesn't necessarily have anything to do with the
start of the function.

To avoid this in heavily-optimised code that lacks many useful source
locations, pick a worst-case "if all else fails" prologue_end location, of
the first instruction that appears to do meaningful computation. It'll be
given the function-scope line number, which should run-on from the start of
the function anyway. This means if your code is completely inverted by the
optimiser, you can at least put a breakpoint at the _start_ like you
expect, even if it's difficult to then step through.

OCHyams commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -2131,36 +2141,121 @@ void DwarfDebug::beginInstruction(const MachineInstr *MI) {
-  return std::make_pair(LineZeroLoc, IsEmptyPrologue);
+
+  // We didn't find a "good" prologue_end, so consider backup locations.
+  // Was there an empty-line location? Return that, and it'll have the
+  // scope-line "flow" into it when it becomes the prologue_end position.
```
I think this could be slightly more clear, something like: "If we found a line-zero location, use that. The prologue_end entry always gets assigned the scope line."


Wait, but maybe I'm wrong - if `IsEmptyPrologue` is true then wouldn't the line zero get used (I'm looking in `emitInitialLocDirective` - non-null line-zero `PrologEndLoc` and with `IsEmptyPrologue` true)? Is this case handled correctly by your patch?

jmorse replied:
(Hoisting to outer discussion as github has lost this review comment in the diff-display),

OCHyams commented on llvm/test/DebugInfo/MIR/X86/dbg-prologue-backup-loc2.mir:
```diff
@@ -0,0 +1,134 @@
+#
+## When picking a "backup" location of the first non-trivial instruction in
+## a function, don't select a location outside of the entry block. We have to
+## give it the functions scope-line, and installing that outside of the entry
```
nit: functions -> function's

OCHyams reviewed (commented):
Looking at the code, I've left some nits and one more meaningful question.

Behaviourally, this SGTM, but I think it would be great if the proposed behaviour also got a quick look from an LLDB-person (@adrian-prantl  / @JDevlieghere ) and/or @dwblaikie, as I could imagine debuggers having special-casing for idiosyncratic compiler output for this kind of thing.

jmorse commented:
(Taking a stab at usernames) CC @tedwoodward from the conference, I believe I mentioned this patch as a revising-the-placement-of-prologue_end thing that was coming along

jmorse committed:
Review feedback

jmorse commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -2131,36 +2141,121 @@ void DwarfDebug::beginInstruction(const MachineInstr *MI) {
+      // Bail if we're also at the end of the function.
+      if (NextBBIter == MF->end())
+        break;
+      CurBlock = NextBBIter;
```
Done

jmorse commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -2131,36 +2141,121 @@ void DwarfDebug::beginInstruction(const MachineInstr *MI) {
+      CurInst = NextBBIter->begin();
+    } else {
+      // Continue examining the current block.
+      CurInst = NextInst;
```
SGTM

jmorse commented:
(Hoisting to here as github has lost the inner comments),

The interaction with linezero prologue_ends is interesting -- I don't think there's anything useful that comes from putting prologue_end on line zero as it doesn't communicate anything further about the function: the frame is setup, but there's no particular line you've stopped at. I think the code selecting for it was part of the previous strategy of "Hang prologue_end onto another line-table entry".

With the revision I've removed the prioritisation of line-zero locations over the "first meaningful instruction" location, which independently makes sense IMO. I've also added a filter to not set prologue_end on the first meaningful instruction if it's also line-zero: it won't communicate anything useful, and giving it a real line-number would be misleading. The outcome is then no prologue_end flag being set, which is the most accurate outcome IMO in these situations.

In terms of test coverage this means `CodeGen/X86/no-non-zero-debug-loc-prologue.ll` no longer generates a prologue, which I've added an implicit-check-not to test for.

OCHyams commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -2192,12 +2279,20 @@ DwarfDebug::emitInitialLocDirective(const MachineFunction &MF, unsigned CUID) {
+    // place prologue_end.
+    if (PrologEndLoc) {
+      const DebugLoc &DL = PrologEndLoc->getDebugLoc();
+      if (!DL || DL->getLine() != 0)
```
I feel slightly confused reading the comment and the current `if` expression. I read the comment as saying "... Thus, only early-return with this location if it's a valid location", but `!DL` is not a location we can use?

Should this be `if (DL && DL->getLine() != 0)` instead of `if (!DL || DL->getLine() != 0)`?

jmorse replied:
I guess it's not clear that the real purpose is simply filtering out line-zero-locations -- those with an _empty_ DebugLoc get given the scope line due to an earlier patch. (It's unfortunate that we've got "empty" nullptr DebugLocs, but also "empty" source locations in the form of line-zero, but here we are).

Without this awkward logic we would end up with zero-length linetable entries for the scope line, before then having a prologue_end linetable entry.

OCHyams replied:
Oh right, yep I see that now, thanks. This all feels a bit tangled up in an unfortunate way, but I've not got any immediate ideas (possibly there's some way of refactoring this, but maybe not), so LGTM.

OCHyams reviewed (commented):
I have another inline question, just to clear up some potential confusion on my part. Otherwise LGTM & I'm happy to _Accept_ once that's resolved.

jmorse committed:
Reword some comments

OCHyams reviewed (approved):
LGTM

jmorse committed:
Merge remote-tracking branch 'github/main' into HEAD

Conflicts:
	llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
	llvm/test/DebugInfo/MIR/X86/empty-inline.mir

jmorse commented:
Pushed up a merge to master to resolve conflicts:
 * In DwarfDebug.cpp some extra basic-block-sections logic has landed. I've placed the logic for "put prologue_end here" in `beginInstruction` before the basic-block-sections logic, as I believe that's guarding logic that considers what the previous linetable entry is.
 * empty-inline.mir has received some updates in the meantime; the diff from 'main' is now that prologue_end shifts to the first instruction with the scope-line, instead of appearing after some loads and branching. (Which is the whole point of this patch)

I believe these are all obvious fixes, will leave the patch here for a few hours to use the CI befor emerging.

Michael137 commented:
Looks like this is breaking the `source-manager/TestSourceManager.py` LLDB test on macOS: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/15127/execution/node/97/log/
```
======================================================================
FAIL: test_artificial_source_location (TestSourceManager.SourceManagerTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/source-manager/TestSourceManager.py", line 325, in test_artificial_source_location
    self.expect(
  File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/packages/Python/lldbsuite/test/lldbtest.py", line 2475, in expect
    self.fail(log_msg)
AssertionError: Ran command:
"process status"

Got output:
Process 99092 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x000000010402bf8c a.out`main at artificial_location.c:3
   1   	int foo() { return 42; }
   2   	
-> 3   	int main() {
   4   	#line 0
   5   	  return foo();
   6   	}

Expecting sub string: "stop reason = breakpoint" (was found)
Expecting sub string: "artificial_location.c:0" (was not found)
Config=arm64-/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang
----------------------------------------------------------------------
```

Could you please take a look?

DavidSpickett commented:
Same on Arm and AArch64 - https://lab.llvm.org/buildbot/#/builders/59/builds/8032.

We expect to be stopped at line 0 `artificial_location.c:0` but are instead at line 3 `artificial_location.c:3`. Which could be the intended change of this patch, not sure at a glance.

jmorse commented:
(Ah, I see the buildkite had another 8Mb of test outputs that it wasn't showing me, guh),

Thanks for the notice; I feel like this is exactly the intended outcome of this change, i.e. ensuring that we produce a steppable location at the start of degenerate functions (including those with #line0 !). That being said, I can't replicate this perfectly locally, I'll work on that, more data in an hour or so.

I can revert if that's the easiest; it's sort of my intention that that test will become redundant though, as the compiler Should (TM) be able to describe the behaviour that LLDB is emulating in the linetable most of the time.

jmorse commented:
Ooooo, and there's a duplicate line-entry added as well, that's a good enough reason to revert too; sorry for the bother.

jmorse commented:
For when I get back to this tomorrow -- is the intention of that test (artificial_location in TestSourceManager) specifically about stepping into prologues + the start of a function, or is it about stimulating LLDBs handling of line-zero locations? I can almost certainly cook up something that stops on line-zero somehow (currently thinking a crash).

labath commented:
It's the latter (the test checks lldb's output when it stops on line zero), and it just so happens that I have a patch (#115876) up which modifies that test. I wasn't able to get it to do what I wanted using the approach in that test, so I implemented something hopefully more robust (it explicitly finds an instruction with line==0 and sets a breakpoint there). I think that should be compatible your changes here.

hokein commented:
This patch introduces an undefined behavior in clang:

```
$ cat /t/t.ii
int a() {}
$ ./bin/clang -O2 -g -c /t/t.ii 
/t/t.ii:1:10: warning: non-void function does not return a value [-Wreturn-type]
    1 | int a() {}
      |          ^
/workspace/llvm-project/llvm/include/llvm/ADT/ilist_iterator.h:169:12: runtime error: reference binding to null pointer of type 'const llvm::MachineInstr'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior/workspace/llvm-project/llvm/include/llvm/ADT/ilist_iterator.h:169:12 in 
```

jmorse commented:
@labath much appreciated, I've got a rebased reapplication in #116084 where I'll use the CI to check this passes,

@hokein would there be any further context than that -- presumably UBSan can produce a stack trace? There are a lot of references to `MachineInstr` here, and not a lot to go on.

jmorse commented:
Rebased patch passes the online CI, and a ubsan stage2 build of clang completed without incident (assuming I've built it correctly), and didn't fire on the given reproducer. Possibly it's a case of mistaken commit identity, or composition with something else that landed at the time?

I'll re-land as the reported issues seem clean to me -- @hokein please do let me know if this reoccurs.

hokein commented:
> @hokein would there be any further context than that -- presumably UBSan can produce a stack trace? There are a lot of references to `MachineInstr` here, and not a lot to go on.

Unfortunately, usan-built clang didn't give a stacktrace. I see a normal assertion-enabled clang (built on bf483ddb4206) also crashes, stacktrace:

```
$ ./bin/clang -O2 -g -c /t/t.ii                                                                                                                                                                                                                   <<<
/t/t.ii:1:10: warning: non-void function does not return a value [-Wreturn-type]
    1 | int a() {}
      |          ^
clang: /workspace/llvm-project/llvm/include/llvm/ADT/ilist_iterator.h:168: reference llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::MachineBasicBlock, true, false, void, false, void>, false, true>::operator*() const [OptionsT = llvm::ilist_detail::node_options<llvm::MachineBasicBlock, true, false, void, false, void>, IsReverse = false, IsConst = true]: Assertion `!NodePtr->isKnownSentinel()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: ./bin/clang -O2 -g -c /t/t.ii
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module '/t/t.ii'.
4.      Running pass 'X86 Assembly Printer' on function '@_Z1av'
 #0 0x000055848fd8e988 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /workspace/llvm-project/llvm/lib/Support/Unix/Signals.inc:723:13
 #1 0x000055848fd8ca90 llvm::sys::RunSignalHandlers() /workspace/llvm-project/llvm/lib/Support/Signals.cpp:106:18
 #2 0x000055848fd08326 (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) /workspace/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:73:5
 #3 0x000055848fd08326 CrashRecoverySignalHandler(int) /workspace/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:390:51
 #4 0x00007f09fac591a0 (/lib/x86_64-linux-gnu/libc.so.6+0x3d1a0)
 #5 0x00007f09faca70ec __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #6 0x00007f09fac59102 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #7 0x00007f09fac424f2 abort ./stdlib/abort.c:81:7
 #8 0x00007f09fac42415 _nl_load_domain ./intl/loadmsgcat.c:1177:9
 #9 0x00007f09fac51d32 (/lib/x86_64-linux-gnu/libc.so.6+0x35d32)
#10 0x0000558490a8ec98 findPrologueEndLoc(llvm::MachineFunction const*) /workspace/llvm-project/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:0:32
#11 0x0000558490a8ec98 llvm::DwarfDebug::emitInitialLocDirective(llvm::MachineFunction const&, unsigned int) /workspace/llvm-project/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:2279:53
#12 0x0000558490a8f0c2 llvm::DwarfDebug::beginFunctionImpl(llvm::MachineFunction const*) /workspace/llvm-project/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:2327:16
#13 0x0000558490a61b0e llvm::AsmPrinter::emitFunctionHeader() /workspace/llvm-project/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:1057:37
#14 0x0000558490a645cd llvm::AsmPrinter::emitFunctionBody() /workspace/llvm-project/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:1768:3
#15 0x000055848ebddaf1 llvm::X86AsmPrinter::runOnMachineFunction(llvm::MachineFunction&) /workspace/llvm-project/llvm/lib/Target/X86/X86AsmPrinter.cpp:91:3
#16 0x000055848f3788f6 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /workspace/llvm-project/llvm/lib/CodeGen/MachineFunctionPass.cpp:94:13
#17 0x000055848f817ae6 llvm::FPPassManager::runOnFunction(llvm::Function&) /workspace/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1406:27
#18 0x000055848f81e432 llvm::FPPassManager::runOnModule(llvm::Module&) /workspace/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1452:13
#19 0x000055848f818255 (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /workspace/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1521:27
#20 0x000055848f818255 llvm::legacy::PassManagerImpl::run(llvm::Module&) /workspace/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:539:44
#21 0x000055848ff7ea6e llvm::TimeTraceScope::~TimeTraceScope() /workspace/llvm-project/llvm/include/llvm/Support/TimeProfiler.h:206:9
#22 0x000055848ff7ea6e (anonymous namespace)::EmitAssemblyHelper::RunCodegenPipeline(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>&, std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile>>&) /workspace/llvm-project/clang/lib/CodeGen/BackendUtil.cpp:1202:3
#23 0x000055848ff7ea6e (anonymous namespace)::EmitAssemblyHelper::EmitAssembly(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) /workspace/llvm-project/clang/lib/CodeGen/BackendUtil.cpp:1224:3
#24 0x000055848ff7ea6e clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) /workspace/llvm-project/clang/lib/CodeGen/BackendUtil.cpp:1387:13
#25 0x00005584904f8c00 std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>::~unique_ptr() /usr/bin/../lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/unique_ptr.h:397:6
#26 0x00005584904f8c00 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) /workspace/llvm-project/clang/lib/CodeGen/CodeGenAction.cpp:354:3

```

jmorse commented:
I've managed to replicate it locally now (I must have been doing something wrong before), should be an easy fix once I've pinned it down, hopefully imminently.

jmorse commented:
Looks like this is a proper problem when there's no body to a function (not something I'd considered possible) -- I'll add a filter to `emitInitialLocDirective` once it's tested.

mikaelholmen commented:
> > @hokein would there be any further context than that -- presumably UBSan can produce a stack trace? There are a lot of references to `MachineInstr` here, and not a lot to go on.
> 
> Unfortunately, usan-built clang didn't give a stacktrace. I see a normal assertion-enabled clang (built on [bf483dd](https://github.com/llvm/llvm-project/commit/bf483ddb42065405e345393e022dc72357ec5a3a)) also crashes, stacktrace:
> 
> ```
> $ ./bin/clang -O2 -g -c /t/t.ii                                                                                                                                                                                                                   <<<
> /t/t.ii:1:10: warning: non-void function does not return a value [-Wreturn-type]
>     1 | int a() {}
>       |          ^
> clang: /workspace/llvm-project/llvm/include/llvm/ADT/ilist_iterator.h:168: reference llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::MachineBasicBlock, true, false, void, false, void>, false, true>::operator*() const [OptionsT = llvm::ilist_detail::node_options<llvm::MachineBasicBlock, true, false, void, false, void>, IsReverse = false, IsConst = true]: Assertion `!NodePtr->isKnownSentinel()' failed.
> PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
> Stack dump:
> 0.      Program arguments: ./bin/clang -O2 -g -c /t/t.ii
> 1.      <eof> parser at end of file
> 2.      Code generation
> 3.      Running pass 'Function Pass Manager' on module '/t/t.ii'.
> 4.      Running pass 'X86 Assembly Printer' on function '@_Z1av'
>  #0 0x000055848fd8e988 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /workspace/llvm-project/llvm/lib/Support/Unix/Signals.inc:723:13
>  #1 0x000055848fd8ca90 llvm::sys::RunSignalHandlers() /workspace/llvm-project/llvm/lib/Support/Signals.cpp:106:18
>  #2 0x000055848fd08326 (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) /workspace/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:73:5
>  #3 0x000055848fd08326 CrashRecoverySignalHandler(int) /workspace/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:390:51
>  #4 0x00007f09fac591a0 (/lib/x86_64-linux-gnu/libc.so.6+0x3d1a0)
>  #5 0x00007f09faca70ec __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
>  #6 0x00007f09fac59102 gsignal ./signal/../sysdeps/posix/raise.c:27:6
>  #7 0x00007f09fac424f2 abort ./stdlib/abort.c:81:7
>  #8 0x00007f09fac42415 _nl_load_domain ./intl/loadmsgcat.c:1177:9
>  #9 0x00007f09fac51d32 (/lib/x86_64-linux-gnu/libc.so.6+0x35d32)
> #10 0x0000558490a8ec98 findPrologueEndLoc(llvm::MachineFunction const*) /workspace/llvm-project/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:0:32
> #11 0x0000558490a8ec98 llvm::DwarfDebug::emitInitialLocDirective(llvm::MachineFunction const&, unsigned int) /workspace/llvm-project/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:2279:53
> #12 0x0000558490a8f0c2 llvm::DwarfDebug::beginFunctionImpl(llvm::MachineFunction const*) /workspace/llvm-project/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:2327:16
> #13 0x0000558490a61b0e llvm::AsmPrinter::emitFunctionHeader() /workspace/llvm-project/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:1057:37
> #14 0x0000558490a645cd llvm::AsmPrinter::emitFunctionBody() /workspace/llvm-project/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:1768:3
> #15 0x000055848ebddaf1 llvm::X86AsmPrinter::runOnMachineFunction(llvm::MachineFunction&) /workspace/llvm-project/llvm/lib/Target/X86/X86AsmPrinter.cpp:91:3
> #16 0x000055848f3788f6 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /workspace/llvm-project/llvm/lib/CodeGen/MachineFunctionPass.cpp:94:13
> #17 0x000055848f817ae6 llvm::FPPassManager::runOnFunction(llvm::Function&) /workspace/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1406:27
> #18 0x000055848f81e432 llvm::FPPassManager::runOnModule(llvm::Module&) /workspace/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1452:13
> #19 0x000055848f818255 (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /workspace/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1521:27
> #20 0x000055848f818255 llvm::legacy::PassManagerImpl::run(llvm::Module&) /workspace/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:539:44
> #21 0x000055848ff7ea6e llvm::TimeTraceScope::~TimeTraceScope() /workspace/llvm-project/llvm/include/llvm/Support/TimeProfiler.h:206:9
> #22 0x000055848ff7ea6e (anonymous namespace)::EmitAssemblyHelper::RunCodegenPipeline(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>&, std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile>>&) /workspace/llvm-project/clang/lib/CodeGen/BackendUtil.cpp:1202:3
> #23 0x000055848ff7ea6e (anonymous namespace)::EmitAssemblyHelper::EmitAssembly(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) /workspace/llvm-project/clang/lib/CodeGen/BackendUtil.cpp:1224:3
> #24 0x000055848ff7ea6e clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) /workspace/llvm-project/clang/lib/CodeGen/BackendUtil.cpp:1387:13
> #25 0x00005584904f8c00 std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>::~unique_ptr() /usr/bin/../lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/unique_ptr.h:397:6
> #26 0x00005584904f8c00 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) /workspace/llvm-project/clang/lib/CodeGen/CodeGenAction.cpp:354:3
> ```

We see this crash downstream as well. Is a fix far away?

jmorse commented:
Sorry for the bother; fixed in 251958f357. @OCHyams could you do some post-commit review on that?

mikaelholmen commented:
> Sorry for the bother; fixed in [251958f](https://github.com/llvm/llvm-project/commit/251958f3570730f58d1337ac6d00f03ee6a839fe). @OCHyams could you do some post-commit review on that?

Thank you! I've verified that the fix solves the problems we saw.

OCHyams commented:
> Sorry for the bother; fixed in https://github.com/llvm/llvm-project/commit/251958f3570730f58d1337ac6d00f03ee6a839fe. @OCHyams could you do some post-commit review on that?

LGTM (commented there too)
